lib/Scheduler/bb_spill.cpp

```diff
@@ -847,8 +855,101 @@ FUNC_RESULT BBWithSpill::Enumerate_(Milliseconds startTime,
+      return UpdtOptmlSchedFrstPss(crntSched, node);
+    else
+      return UpdtOptmlSchedScndPss(crntSched, node);
+  }
+
+  else
+    return UpdtOptmlSchedWghtd(crntSched, node);
+}
+
+/*****************************************************************************/
+
+SmallVector<InstCount, 4>
+BBWithSpill::UpdtOptmlSchedFrstPss(InstSchedule *crntSched,
+                                   LengthCostEnumerator *) {
+  InstCount crntCost;
+  InstCount crntExecCost;
+  InstCount TmpSpillCost;
+  llvm::SmallVector<InstCount, 4> retVec;
+
+  //  crntCost = CmputNormCost_(crntSched, CCM_DYNMC, crntExecCost, false);
+  crntCost = CmputNormCost_(crntSched, CCM_STTC, crntExecCost, false);
+
+  //#ifdef IS_DEBUG_SOLN_DETAILS_2
+  Logger::Info(
+      "Found a feasible sched. of length %d, spill cost %d and tot cost %d",
+      crntSched->GetCrntLngth(), crntSched->GetSpillCost(), crntCost);
+  //  crntSched->Print(Logger::GetLogStream(), "New Feasible Schedule");
+  //#endif
+
+  TmpSpillCost =
+      GetSpillCostFunc() == SCF_SLIL ? dynamicSlilLowerBound_ : crntSpillCost_;
+
+  if (TmpSpillCost < getBestSpillCost()) {
+    SetBestCost(crntCost);
+    optmlSpillCost_ = TmpSpillCost;
+    setBestSpillCost(optmlSpillCost_);
+    SetBestSchedLength(crntSched->GetCrntLngth());
+    enumBestSched_->Copy(crntSched);
+    bestSched_ = enumBestSched_;
+  }
+
+  retVec.push_back(TmpSpillCost);
+  return retVec;
+}
+
+/*****************************************************************************/
+
+SmallVector<InstCount, 4>
+BBWithSpill::UpdtOptmlSchedScndPss(InstSchedule *crntSched,
+                                   LengthCostEnumerator *) {
+  InstCount crntCost;
+  InstCount crntExecCost;
+  InstCount TmpSpillCost;
+  llvm::SmallVector<InstCount, 4> retVec;
+
+  //  crntCost = CmputNormCost_(crntSched, CCM_DYNMC, crntExecCost, false);
+  crntCost = CmputNormCost_(crntSched, CCM_STTC, crntExecCost, false);
+
+  //#ifdef IS_DEBUG_SOLN_DETAILS_2
+  Logger::Info(
+      "Found a feasible sched. of length %d, spill cost %d and tot cost %d",
+      crntSched->GetCrntLngth(), crntSched->GetSpillCost(), crntCost);
+  //  crntSched->Print(Logger::GetLogStream(), "New Feasible Schedule");
+  //#endif
+
+  TmpSpillCost =
+      GetSpillCostFunc() == SCF_SLIL ? dynamicSlilLowerBound_ : crntSpillCost_;
+
+  if (TmpSpillCost <= getSpillCostConstraint()) {
+    SetBestCost(crntCost);
+    optmlSpillCost_ = TmpSpillCost;
+    setBestSpillCost(optmlSpillCost_);
+    SetBestSchedLength(crntSched->GetCrntLngth());
+    enumBestSched_->Copy(crntSched);
+    bestSched_ = enumBestSched_;
+  }
+
+  retVec.push_back(TmpSpillCost);
+  retVec.push_back(crntSched->GetCrntLngth());
+  return retVec;
+}
+
+/*****************************************************************************/
+
+SmallVector<InstCount, 4>
+BBWithSpill::UpdtOptmlSchedWghtd(InstSchedule *crntSched,
+                                 LengthCostEnumerator *) {
+  llvm::SmallVector<InstCount, 4> retVec;
+
```

Comment:
There's a lot of duplicated code between these three functions. Maybe pull out the duplicated code pieces into common functions?